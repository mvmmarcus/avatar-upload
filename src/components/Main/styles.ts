import styled, { css } from "styled-components";

export const Wrapper = styled.main`
  ${({ theme }) => css`
    display: flex;
    justify-content: center;
    align-items: center;
    min-height: 100vh;
    padding: ${theme.spacings.xsmall};
  `}
`;
