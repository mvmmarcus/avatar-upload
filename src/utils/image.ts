type CropImageProps = {
  imageUrl: string;
  cropWidth?: number;
  preserveAspectRatio?: boolean;
  previewImageRatio?: number;
};

const addImageProcess = async (
  imageSource: string
): Promise<HTMLImageElement> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = imageSource;
  });
};

export const cropImage = async ({
  imageUrl,
  cropWidth = 300,
  preserveAspectRatio = true,
  previewImageRatio = 1,
}: CropImageProps) => {
  const image = await addImageProcess(imageUrl);

  const ratio = image.height / image.width;

  const canvas = document.createElement("canvas");
  const context = canvas.getContext("2d");

  if (preserveAspectRatio) {
    canvas.width = image.width < cropWidth ? image.width : cropWidth;
    canvas.height = image.width < cropWidth ? image.height : cropWidth * ratio;

    context?.drawImage(image, 0, 0, canvas.width, canvas.height);
  } else {
    canvas.width = previewImageRatio * image.width;
    canvas.height = canvas.width;

    const cropDistanceX = (image.width - canvas.width) / 2;
    const cropDistanceY = (image.height - canvas.height) / 2;

    context?.drawImage(
      image,
      cropDistanceX,
      cropDistanceY,
      canvas.width,
      canvas.height,
      0,
      0,
      canvas.width,
      canvas.height
    );
  }

  const croppedImageUrl = canvas.toDataURL();

  return croppedImageUrl;
};

export const uploadImage = (file: File): Promise<ProgressEvent<FileReader>> => {
  return new Promise((resolve, reject) => {
    const fileReader = new FileReader();

    fileReader.onload = (event) => resolve(event);
    fileReader.onerror = reject;
    fileReader.readAsDataURL(file);
  });
};
