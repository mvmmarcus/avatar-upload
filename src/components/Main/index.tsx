import React from "react";
import { Wrapper } from "./styles";

type MainProps = {
  children: React.ReactNode;
};

const Main = ({ children }: MainProps) => {
  return <Wrapper>{children}</Wrapper>;
};

export default Main;
