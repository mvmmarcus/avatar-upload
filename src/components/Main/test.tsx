import React from "react";

import { screen } from "@testing-library/react";

import Main from ".";
import { renderWithTheme } from "utils/test-utils";

describe("<Main />", () => {
  it("should render main component", () => {
    const { container } = renderWithTheme(<Main>Main</Main>);

    expect(screen.getByText(/main/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
