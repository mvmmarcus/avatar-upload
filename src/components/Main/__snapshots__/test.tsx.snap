// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should render main component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 100vh;
  padding: 1.6rem;
}

<main
  class="c0"
>
  Main
</main>
`;
